Sources/NIOHTTP1/HTTPTypes.swift

```diff
@@ -472,7 +472,108 @@ public struct HTTPHeaders: CustomStringConvertible, ExpressibleByDictionaryLiter
+        var values: [Substring] = []
+        for result in result {
+            var parser = HTTPHeaderValueParser(result)
+            while let value = parser.nextValue() {
+                values.append(value)
+            }
+        }
+        return values
+    }
+}
+
+private struct HTTPHeaderValueParser {
+    var current: Substring
+
+    init(_ string: String) {
+        self.current = .init(string)
+    }
+
+    mutating func nextValue() -> Substring? {
+        self.popWhitespace()
+        guard !self.current.isEmpty else {
+            return nil
+        }
+        let value: Substring
+        if self.current.first == "\"" {
+            self.pop()
+            guard let nextDoubleQuote = self.nextUnescapedDoubleQuote() else {
+                return nil
+            }
+            value = self.pop(to: nextDoubleQuote).unescapingDoubleQuotes()
+            self.pop()
+            if let comma = self.nextComma() {
+                _ = self.pop(to: comma)
+                self.pop()
+            }
+        } else if let comma = self.nextComma() {
+            value = self.pop(to: comma)
+            self.pop()
+        } else {
+            value = self.pop(to: self.current.endIndex).trimWhitespace()
+        }
+        return value
+    }
+
+    private func nextUnescapedDoubleQuote() -> Substring.Index? {
+        var startIndex = self.current.startIndex
+        var nextDoubleQuote: Substring.Index?
+        while nextDoubleQuote == nil {
+            guard let possibleDoubleQuote = self.current[startIndex...].firstIndex(of: "\"") else {
+                return nil
+            }
+            // Check if quote is escaped.
+            if self.current.startIndex == possibleDoubleQuote || self.current[self.current.index(before: possibleDoubleQuote)] != "\\" {
+                nextDoubleQuote = possibleDoubleQuote
+            } else if possibleDoubleQuote < self.current.endIndex {
+                startIndex = self.current.index(after: possibleDoubleQuote)
+            } else {
+                return nil
+            }
+        }
+        return nextDoubleQuote
+    }
+
+    private func nextComma() -> Substring.Index? {
+        return self.current.firstIndex(of: ",")
+    }
+
+    private mutating func popWhitespace() {
+        if let nonWhitespace = self.current.firstIndex(where: { !$0.isLinearWhitespace }) {
+            self.current = self.current[nonWhitespace...]
+        } else {
+            self.current = ""
+        }
+    }
+
+    private mutating func pop() {
+        self.current = self.current.dropFirst()
+    }
+
+    private mutating func pop(to index: Substring.Index) -> Substring {
+        let value = self.current[..<index]
+        self.current = self.current[index...]
+        return value
+    }
+}
+
+private extension Substring {
+    /// Converts all `\"` to `"`.
+    func unescapingDoubleQuotes() -> Substring {
+        return self.split(separator: "\\").lazy.reduce(into: "") { (result, part) in
```

Comment:
`self.lazy.split.reduce` seems to work fine for me, and it's the correct spelling.

Reply:
OK, cool. Thanks. I've taken this PR out of draft.

Reply:
Sorry, I think this meant moving the `lazy` to the left.

Reply:
Oh sorry, I can't read. Lemme try that again.

Reply:
Ah okay, I found the problem. The lazy subsequence's `CustomStringConvertible` conformance was different. I fixed this by using concatenation instead.